api/src/main/java/org/apache/iceberg/types/Types.java

```diff
@@ -48,6 +48,8 @@ private Types() {}
           .put(TimeType.get().toString(), TimeType.get())
           .put(TimestampType.withZone().toString(), TimestampType.withZone())
           .put(TimestampType.withoutZone().toString(), TimestampType.withoutZone())
+          .put(TimestampnsType.withZone().toString(), TimestampnsType.withZone())
```

Comment:
General point: Should this be a new type, or should Timestamp instead have a precision field?

One thing that would speak for the latter is that with /without time zone is also not an own type but instead a field of timestamp.

Reply:
@JFinis we discussed `timestamp[MILLIS|MICROS|NANOS]` in [the proposal/design doc](https://docs.google.com/document/d/1bE1DcEGNzZAMiVJSZ0X1wElKLNkT9kRkk0hDlfkXzvU/edit), but decided against it.

Reply:
@jacobmarble I don't see any discussion or any rationale in the document.

Reply:
One could argue that so far the design was to separate different timestamps through fields, not new types. This deviates from the design so far. I would like to understand the rationale for this deviation.

Reply:
@JFinis there are resolved comments in the doc, visible without being the document owner, but you'll have to click the "open comment history" button to see them.

The document history is not visible to anyone but the owner, so I've attached PDFs at two stages: my original draft (7-sep-2023) and the doc as understood at the last community sync (19-sep-2023).

[At the very end of the last community sync](https://youtu.be/MIreG41AabI?t=3300), I understood that we had enough consensus to start writing code.

[Nanosecond Timestamps in Apache Iceberg 7-sep-2023.pdf](https://github.com/apache/iceberg/files/12772864/Nanosecond.Timestamps.in.Apache.Iceberg.7-sep-2023.pdf)
[Nanosecond Timestamps in Apache Iceberg 19-sep-2023.pdf](https://github.com/apache/iceberg/files/12772865/Nanosecond.Timestamps.in.Apache.Iceberg.19-sep-2023.pdf)

> the design was to separate different timestamps through fields, not new types

Help me out, where do you read that?

Reply:
Thanks, now I see the discussions! That being said, I don't see a discussion that discusses whether a new type should be used vs. a new field in the timestamp type. Can you point me to that discussion? Maybe I'm missing the forest for the trees here. Or the discussion was removed due to the text to which it refers to being removed?

I see that it was a suggestion in your old versions. Therefore, I would like to understand the rationale of the person who was against it.

> > the design was to separate different timestamps through fields, not new types

> Help me out, where do you read that?

I read that in the code itself: There is only one `timestamp` type with an `adjust-to-utc` flag. Instead, there could have been two types, `timestamp` and `timestamptz`. Thus, the person doing the initial design decided for a flag instead of two different types. 

And now you add another dimension in which timestamps can be different. Contrarily, the design closest to the initial design would be to have yet another (enum) field on the timestamp type instead of a new type. This would also help with the clumsy backward compatible naming of `timestamp` (without any suffix) and now `timestamp_ns`, where one type has a suffix while the other one hasn't. With a new field, everything would just stay `timestamp` and the field would implicitly be Millis, if it's not set.

So at least for me, this points towards having the field. Though, maybe there is a disadvantage of the field approach I don't see here.

Reply:
> I don't see a discussion that discusses whether a new type should be used vs. a new field in the timestamp type

> There is only one timestamp type with an adjust-to-utc flag. Instead, there could have been two types, timestamp and timestamptz.

> clumsy backward compatible naming of timestamp (without any suffix) and now timestamp_ns, where one type has a suffix while the other one hasn't.

> maybe there is a disadvantage of the field approach I don't see here.

We may be talking past each other wrt the Iceberg types vs the Java implementation of those types.

- The design considers **existing Iceberg types** `timestamp` and `timestamptz` (microsecond units), then **adds new types** `timestamp_ns` and `timestamptz_ns` (nanosecond units).
  - @rdblue guided the design decision away from (1) parameterized type names `timestamp[MILLIS|MICROS|NANOS]` and (2) adding milliseconds at the same time we add nanoseconds.

- This proposal for Java implementation of the design considers existing type `TimestampType` (including field `private final boolean adjustToUTC`), then adds new type `TimestampnsType` (including field `private final boolean adjustToUTC`).
  - The concern @JFinis raises is that Java `TimestampType` already represents two Iceberg types, so it would make sense that the 
  - Do I understand you correctly?
  - If so: I did start this change by extending Java type `TimestampType` with a new field, but backed out and started over with a new Java type. I was overwhelmed by the variety of cases where the micro/nano branch would be needed. In contrast, adding a new type made it easier for me to self-audit and test, and removed any confusion to a user of the API. The approach could certainly be revisited.

>  With a new field, everything would just stay timestamp and the field would implicitly be Millis, if it's not set.

nit: Currently, `timestamp` and `timestamptz` represent microseconds, not milliseconds. And yes, I had proposed that `timestamp` continue to refer to microseconds for backward compatibility.

Reply:
@JFinis PTAL

It was decided [in the spec PR](https://github.com/apache/iceberg/pull/8683#issuecomment-1749792874) that we should add milliseconds, after all.

In this PR, I've split the work into three commits:
- add class `Timestampns`
- add class `Timestampms`
- remove the above classes, replace them with a new field in `Timestamp`